# HG changeset patch
# User Jason Kim <jasonwkim@google.com>
# Parent 4a6a667b4f2542b4bdca4e67b0bc8e9e35cd78aa
upstream: reorg isBlockOnlyReachableByFallthough() to have a helper routine

diff -r 4a6a667b4f25 lib/CodeGen/AsmPrinter/AsmPrinter.cpp
--- a/lib/CodeGen/AsmPrinter/AsmPrinter.cpp	Wed Feb 02 19:05:20 2011 -0800
+++ b/lib/CodeGen/AsmPrinter/AsmPrinter.cpp	Thu Aug 11 23:33:48 2011 -0700
@@ -1955,6 +1955,28 @@
     OutStreamer.EmitSymbolAttribute(Sym, Attr);
 }
 
+// CheckIsMIFallThrough -
+// This is a helper routine for isBlockOnlyReachableByFallthough
+static bool CheckIsMIFallThrough(MachineBasicBlock::iterator II,
+                                 const MachineBasicBlock *MBB) {
+  MachineInstr &MI = *II;
+  // If it is not a simple branch, we are in a table somewhere.
+  if (!MI.getDesc().isBranch() || MI.getDesc().isIndirectBranch())
+    return false;
+
+  // If we are the operands of one of the branches, this is not
+  // a fall through.
+  for (MachineInstr::mop_iterator OI = MI.operands_begin(),
+         OE = MI.operands_end(); OI != OE; ++OI) {
+    const MachineOperand& OP = *OI;
+    if (OP.isJTI())
+      return false;
+    if (OP.isMBB() && OP.getMBB() == MBB)
+      return false;
+  }
+  return true;
+}
+
 /// isBlockOnlyReachableByFallthough - Return true if the basic block has
 /// exactly one predecessor and the control transfer mechanism between
 /// the predecessor and this block is a fall-through.
@@ -1984,22 +2006,8 @@
   // Check the terminators in the previous blocks
   for (MachineBasicBlock::iterator II = Pred->getFirstTerminator(),
          IE = Pred->end(); II != IE; ++II) {
-    MachineInstr &MI = *II;
-
-    // If it is not a simple branch, we are in a table somewhere.
-    if (!MI.getDesc().isBranch() || MI.getDesc().isIndirectBranch())
+    if (!CheckIsMIFallThrough(II, MBB))
       return false;
-
-    // If we are the operands of one of the branches, this is not
-    // a fall through.
-    for (MachineInstr::mop_iterator OI = MI.operands_begin(),
-           OE = MI.operands_end(); OI != OE; ++OI) {
-      const MachineOperand& OP = *OI;
-      if (OP.isJTI())
-        return false;
-      if (OP.isMBB() && OP.getMBB() == MBB)
-        return false;
-    }
   }
 
   return true;
